ical/tzif/tz_rule.py

```diff
@@ -200,64 +181,98 @@ def default_dst_offset(cls, values: dict[str, Any]) -> dict[str, Any]:
+
+        # make sure we have reached the end of the string
+        if self._pos < len(self._str):
+            raise ValueError(
+                f"Unable to parse TZ string: {self._str}. Unexpected end of string at position {self._pos}."
+            )
+
+        return Rule(std=std, dst=dst, dst_start=std_start, dst_end=std_end)
+
+    def _parse_tz_rule_occurrence(self) -> RuleOccurrence | None:
+        """Parse the TZ string into a RuleOccurrence object."""
+        match = _OFFSET_RE_PATTERN.match(self._str, self._pos)
+        if not match:
+            return None
+
+        self._pos = match.end()
+
+        return RuleOccurrence(
+            name=match.group("name"), offset=_parse_time(match.groupdict()) or _ZERO
         )
+
+    def _parse_tz_rule_date(self) -> Union[RuleDate, RuleDay, None]:
+        """Parse the TZ string into a RuleDate or RuleDay object."""
+        match = _START_END_RE_PATTERN.match(self._str, self._pos)
+        if not match:
+            return None
+
+        self._pos = match.end()
+
+        if match["day_of_year"] is not None:
+            return RuleDay(
+                day_of_year=int(match.group("day_of_year")),
+                time=_parse_time(match.groupdict()) or _DEFAULT_TIME_DELTA,
+            )
+
+        return RuleDate(
+            month=int(match.group("month")),
+            week_of_month=int(match.group("week_of_month")),
+            day_of_week=int(match.group("day_of_week")),
+            time=_parse_time(match.groupdict()) or _DEFAULT_TIME_DELTA,
+        )
+
+
+def build_offset_regex() -> re.Pattern[str]:
+    """Create a regular expression for the given TZ string. Alternative implementation to pyparsing."""
+    # name
+    result = r"(?P<name>(\<[+\-]?\d+\>|[a-zA-Z]+))"
+    # offset
+    result += (
+        r"((?P<hour>[+-]?\d+)(?::(?P<minutes>\d{1,2})(?::(?P<seconds>\d{1,2}))?)?)?"
+    )
+    return re.compile(result)
+
+
+def build_start_end_regex() -> re.Pattern[str]:
+    # days in either julian (J prefix) or month.week.day (M prefix) format
+    result = r",(J(?P<day_of_year>\d+)|M(?P<month>\d{1,2})\.(?P<week_of_month>\d)\.(?P<day_of_week>\d))"
+    # time
+    result += (
+        r"(\/(?P<hour>[+-]?\d+)(?::(?P<minutes>\d{1,2})(?::(?P<seconds>\d{1,2}))?)?)?"
     )
-    return tz_rule
+    return re.compile(result)
 
 
 def parse_tz_rule(tz_str: str) -> Rule:
     """Parse the TZ string into a Rule object."""
+    parser = TzParser(tz_str)
+    return parser.parse()
 
-    start_gator = tz_str.startswith("<")
-    is_julian_date = ",J" in tz_str
-    tz_rule = _create_parser(start_gator, is_julian_date)
-    try:
-        result = tz_rule.parse_string(tz_str, parse_all=True)
-    except ParseException as err:
-        raise ValueError(f"Unable to parse TZ string: {tz_str}") from err
 
-    return Rule.parse_obj(result.as_dict())
+_OFFSET_RE_PATTERN = build_offset_regex()
+_START_END_RE_PATTERN = build_start_end_regex()
```

Comment:
lets move these constants and `build_offset_regex` and `build_start_end_regex` above the `TzParser` declaration.